api/base/serializers.py

```diff
@@ -404,9 +404,17 @@ def to_internal_value(self, data):
             type_ = get_meta_type(self.root.child, request)
         else:
             type_ = get_meta_type(self.root, request)
-
-        if type_ != data:
-            raise api_exceptions.Conflict(detail=('This resource has a type of "{}", but you set the json body\'s type field to "{}". You probably need to change the type field to match the resource\'s type.'.format(type_, data)))
+        if StrictVersion(request.version) != StrictVersion('2.15'):
+            if type_ != data:
+                raise api_exceptions.Conflict(detail=('This resource has a type of "{}", but you set the json body\'s type field to "{}". You probably need to change the type field to match the resource\'s type.'.format(type_, data)))
+        else:
+            kebab_case = str(type_).replace('-', '_')
+            if type_ != data:
+                if kebab_case == data:
+                    type_ = kebab_case
+                    self.context['request'].META.setdefault('warning', 'As of API Version 2.15, all types are now Kebab-case. 2.15 will accept snake_case, but this will be deprecated in future versions.')
+                else:
+                    raise api_exceptions.Conflict(detail=('This resource has a type of "{}", but you set the json body\'s type field to "{}". You probably need to change the type field to match the resource\'s type.'.format(type_, data)))
```

Comment:
I don't love how this line is the same as 409 - can you rework your logic to be less repetitive?  Would probably go away with other recommended change here.